### PR TITLE
QOS support

### LINKED ISSUE #439: Support for QOS capabilities
**Issue Type**

- Template enhancement
- Role enhancement

**Is your feature request related to a problem? Please describe.**
To support QOS configuration on interface level and in global mode.

This includes:
- QOS profiles for ports (i.e trust dscp, schedulers)
- DSCP to traffic-class mapping
- Interface service-profile support

This should go on all EOS devices.

**Describe the solution you'd like**

```
qos map dscp {{ VALUE }} to traffic-class {{ CLASS }}
!
qos profile {{ QOS }}
   qos trust dscp
   !
   tx-queue {{ NUMBER }}
      bandwidth percent {{ NUMBER }}
   !
interface Ethernet {{ INTERFACE_NUMBER }}
   service-profile {{ QOS }}
```

**Describe alternatives you've considered**
Static configuration

**Additional context**

### PR DESCRIPTION
**Pull request by onurgashi**
## Change Summary

This is a draft for QOS. Covers QOS profile, QOS mappings, and QOS rewrite.

Would be great to have some feedback on data model. I.e QOS profile and QOS mappings are not in the same location in cfg file.

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [x] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes)
- [ ] Documentation content changes
- [ ] Other (please describe):

## Related Issue(s)

Fixes #439 

## Component(s) name
eos_cli_config_gen
molecule

## How to test

Molecule and HW

## Checklist:

- [x] I have read the [**CONTRIBUTING**](https://www.avd.sh/docs/contributing/) document.
- [x] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] All new and existing tests passed ([`pre-commit`](https://www.avd.sh/docs/installation/development/#python-virtual-environment), `make linting` and `make sanity-lint`).
- [x] I have updated [molecule CI](https://github.com/aristanetworks/ansible-avd/tree/devel/ansible_collections/arista/avd/molecule) testing accordingly

### CODE REVIEW

**Comment by onurgashi** [comment]
Added interface level support. There is a small thing we need to look into with the location of qos profile and qos the rest of the global knobs. Might be we need to refactor into two templates so we can facilitate the change of location in the running-config.

**Comment by carlbuchmann** [comment]
- [ ] update eos_cli_config_gen documentation

**Review by titom73 — DISMISSED** [review]
Seems there is no documentation generated for QOS settings. We should provide at least a basic doc generated for a QOS summary

File: ansible_collections/arista/avd/roles/eos_cli_config_gen/README.md [hunk 1]
```diff
@@ -781,6 +786,10 @@ ethernet_interfaces:
       enable: < true | false >
     mac_security:
       profile: < profile >
+    service_profile: < qos_profile >
+    qos_trust: < dscp defualt(L3) | cos default(L2) >
```
Comment (ClausHolbechArista) [inline ansible_collections/arista/avd/roles/eos_cli_config_gen/README.md]: typo, def**au**lt(L3)

File: ansible_collections/arista/avd/roles/eos_cli_config_gen/README.md [hunk 2]
```diff
@@ -756,6 +757,10 @@ ethernet_interfaces:
     isis_network_point_to_point: < boolean >
     ptp:
       enable: < true | false >
+    service_profile: < qos_profile >
+    qos_trust: < dscp defualt(L3) | cos default(L2) >
```
Comment (ClausHolbechArista) [inline ansible_collections/arista/avd/roles/eos_cli_config_gen/README.md]: typo, def**au**lt(L3)

Comment (ClausHolbechArista) [reply ansible_collections/arista/avd/roles/eos_cli_config_gen/README.md]: These defaults are not AVD defaults are they? We do not need to describe EOS defaults here.

**Comment by onurgashi** [comment]
I have updated the QOS as:
- Reflect the comments
- Separate profiles with general QOS
- Interface level updated for port-channels and ethernet interfaces from qos_feature to qos.feature
- Documentation has been added for global QOS and QOS profiles
- Molecule tests successful

File: ansible_collections/arista/avd/molecule/eos_cli_config_gen/intended/configs/errdisable.cfg [hunk 3]
```diff
@@ -26,12 +26,14 @@ errdisable recovery cause xcvr-unsupported
 errdisable recovery interval 300
 !
 hostname errdisable
+
```
Comment (ClausHolbechArista) [inline ansible_collections/arista/avd/molecule/eos_cli_config_gen/intended/configs/errdisable.cfg]: These extra empty lines should be removed

File: ansible_collections/arista/avd/molecule/eos_cli_config_gen/intended/configs/errdisable.cfg [hunk 4]
```diff
@@ -26,12 +26,14 @@ errdisable recovery cause xcvr-unsupported
...
 !
 no aaa root
 !
 interface Management1
    description oob_management
    vrf MGMT
    ip address 10.73.255.122/24
+
```
Comment (ClausHolbechArista) [inline ansible_collections/arista/avd/molecule/eos_cli_config_gen/intended/configs/errdisable.cfg]: These extra empty lines should be removed

File: ansible_collections/arista/avd/roles/eos_cli_config_gen/templates/eos-intended-config.j2 [hunk 5]
```diff
@@ -172,6 +174,8 @@ hostname {{ inventory_hostname }}
 {% include 'eos/management-security.j2' %}
 {# management ssh #}
 {% include 'eos/management-ssh.j2' %}
+{# management qos #}
```
Comment (ClausHolbechArista) [inline ansible_collections/arista/avd/roles/eos_cli_config_gen/templates/eos-intended-config.j2]: Not management. Just qos

**Review by ClausHolbechArista — COMMENTED** [review]
Small change required re empty lines. I will try to fix.

**Review by ClausHolbechArista — APPROVED** [review]
LGTM

**Pull request merged**